lib/trino-filesystem/src/test/java/io/trino/filesystem/memory/TestMemoryFileSystemCache.java

```diff
@@ -0,0 +1,131 @@
+
+    private MemoryFileSystem delegate;
+    private CacheFileSystem fileSystem;
+    private MemoryFileSystemCache cache;
+    private CacheKeyProvider cacheKeyProvider;
+
+    @BeforeAll
+    void beforeAll()
+    {
+        MemoryFileSystemCacheConfig configuration = new MemoryFileSystemCacheConfig()
+                .setMaxContentLength(DataSize.ofBytes(MAX_CONTENT_LENGTH))
+                .setCacheTTL(new Duration(8, HOURS));
+        delegate = new MemoryFileSystem();
+        cache = new MemoryFileSystemCache(noopTracer(), configuration);
+        cacheKeyProvider = new DefaultCacheKeyProvider();
+        fileSystem = new CacheFileSystem(delegate, cache, cacheKeyProvider);
+    }
+
+    @Override
+    protected boolean isHierarchical()
+    {
+        return false;
+    }
+
+    @Override
+    protected boolean supportsCreateExclusive()
+    {
+        return true;
+    }
+
+    @Override
+    protected TrinoFileSystem getFileSystem()
+    {
+        return fileSystem;
+    }
+
+    @Override
+    protected Location getRootLocation()
+    {
+        return Location.of("memory://");
+    }
+
+    @Override
+    protected void verifyFileSystemIsEmpty()
+    {
+        assertThat(delegate.isEmpty()).isTrue();
+    }
+
+    @Test
+    public void testMaxContentLength()
+            throws IOException
+    {
+        int fileSize = MAX_CONTENT_LENGTH + 200;
+        Location location = writeFile(fileSize);
+        TrinoInputFile inputFile = getFileSystem().newInputFile(location);
+        long largeFileSkippedCount = cache.getLargeFileSkippedCount();
+        try (TrinoInput input = inputFile.newInput()) {
+            input.readTail(fileSize);
+        }
+        assertThat(cache.getLargeFileSkippedCount()).isGreaterThan(largeFileSkippedCount);
+        assertThat(cache.isCached(cacheKeyProvider.getCacheKey(inputFile).orElseThrow())).isFalse();
+        getFileSystem().deleteFile(location);
+
+        fileSize = MAX_CONTENT_LENGTH - 200;
+        location = writeFile(fileSize);
+        inputFile = getFileSystem().newInputFile(location);
+        try (TrinoInput input = inputFile.newInput()) {
+            input.readTail(fileSize);
+        }
+        assertThat(cache.isCached(cacheKeyProvider.getCacheKey(inputFile).orElseThrow())).isTrue();
+        getFileSystem().deleteFile(location);
+    }
+
+    private Location writeFile(int fileSize)
+            throws IOException
+    {
+        Location location = getRootLocation().appendPath("testMaxContentLength-%s".formatted(UUID.randomUUID()));
+        getFileSystem().deleteFile(location);
+        try (OutputStream outputStream = getFileSystem().newOutputFile(location).create()) {
+            byte[] bytes = new byte[8192];
+            Arrays.fill(bytes, (byte) 'a');
+            int count = 0;
+            while (count < fileSize) {
+                outputStream.write(bytes);
+                count += bytes.length;
+            }
+        }
```

Comment:
Does this produce a file of size `ceil(fileSize / 8192) * 8192`?

Reply:
ya, tweaked that to be exact now